atom/browser/ui/webui/pdf_viewer_handler.cc

```diff
@@ -0,0 +1,211 @@
+  std::string header_value;
+  while (headers->EnumerateHeaderLines(&iter, &header_name, &header_value)) {
+    base::Value* existing_value = nullptr;
+    if (result->Get(header_name, &existing_value)) {
+      base::StringValue* existing_string_value =
+          static_cast<base::StringValue*>(existing_value);
+      existing_string_value->GetString()->append(", ").append(header_value);
+    } else {
+      result->SetString(header_name, header_value);
+    }
+  }
+}
+
+void PopulateStreamInfo(base::DictionaryValue* stream_info,
+                        content::StreamInfo* stream,
+                        const std::string& original_url) {
+  auto headers_dict = base::MakeUnique<base::DictionaryValue>();
+  auto stream_url = stream->handle->GetURL().spec();
+  CreateResponseHeadersDictionary(stream->response_headers.get(),
+                                  headers_dict.get());
+  stream_info->SetString("streamURL", stream_url);
+  stream_info->SetString("originalURL", original_url);
+  stream_info->Set("responseHeaders", std::move(headers_dict));
+}
+
+}  // namespace
+
+PdfViewerHandler::PdfViewerHandler(const std::string& src)
+    : stream_(nullptr), original_url_(src) {}
+
+PdfViewerHandler::~PdfViewerHandler() {}
+
+void PdfViewerHandler::SetPdfResourceStream(content::StreamInfo* stream) {
+  stream_ = stream;
+  if (!!initialize_callback_id_.get()) {
+    auto list = base::MakeUnique<base::ListValue>();
+    list->Set(0, std::move(initialize_callback_id_));
+    Initialize(list.get());
+  }
+}
+
+void PdfViewerHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "initialize",
+      base::Bind(&PdfViewerHandler::Initialize, base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getDefaultZoom",
+      base::Bind(&PdfViewerHandler::GetInitialZoom, base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getInitialZoom",
+      base::Bind(&PdfViewerHandler::GetInitialZoom, base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "setZoom",
+      base::Bind(&PdfViewerHandler::SetZoom, base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getStrings",
+      base::Bind(&PdfViewerHandler::GetStrings, base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "reload", base::Bind(&PdfViewerHandler::Reload, base::Unretained(this)));
+}
+
+void PdfViewerHandler::OnJavascriptAllowed() {
+  auto host_zoom_map =
+      content::HostZoomMap::GetForWebContents(web_ui()->GetWebContents());
+  host_zoom_map_subscription_ =
+      host_zoom_map->AddZoomLevelChangedCallback(base::Bind(
+          &PdfViewerHandler::OnZoomLevelChanged, base::Unretained(this)));
+}
+
+void PdfViewerHandler::OnJavascriptDisallowed() {
+  host_zoom_map_subscription_.reset();
+}
+
+void PdfViewerHandler::Initialize(const base::ListValue* args) {
+  CHECK_EQ(1U, args->GetSize());
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+
+  if (stream_) {
```

Comment:
Should we check whether `initialize_callback_id_ ` is `nullptr`? Or make a `CHECK` here.